### PR TITLE
Renaming of functions and types

### PR DESCRIPTION
**Pull request by joehuchette**

### CODE REVIEW

File: doc/refvariable.rst [hunk 1]
```diff
@@ -114,7 +114,7 @@ JuMP supports modeling with `semidefinite variables <https://en.wikipedia.org/wi
 Note in particular the indexing: 1) exactly two index sets must be specified, 2) they must both be unit ranges starting at 1, 3) no bounds can be provided alongside the ``SDP`` tag. If you wish to impose more complex semidefinite constraints on the variables, e.g. :math:`X - I \succeq 0`, you may instead use the ``Symmetric`` tag, along with a semidefinite constraint::
 
     @defVar(m, X[1:n,1:n], Symmetric)
-    @addSDPConstraint(m, X >= eye(n))
+    @addSDConstraint(m, X >= eye(n))
```
Comment (mlubin) [inline doc/refvariable.rst]: Incorrect find/replace

File: NEWS.md [hunk 2]
```diff
@@ -120,7 +120,7 @@ Version 0.8.0 (February 17, 2015)
   * SCS supported for solving second-order conic problems.
   * ``setXXXCallback`` family deprecated in favor of ``addXXXCallback``.
   * Multiple callbacks of the same type can be registered.
-  * Added support for informational callbacks via ``addInfoCallback``.
+  * Added support for informational callbacks via ``addinfocallback``.
```
Comment (mlubin) [inline NEWS.md]: Shouldn't modify old NEWS entries

File: doc/quickstart.rst [hunk 3]
```diff
@@ -71,27 +71,20 @@ Finally, bounds can depend on variable indices::
...
-languages and has no restrictions. The second utilizes Julia's powerful
-metaprogramming features to get excellent performance even for large problems,
-but has some restrictions on how they can be used.
-
-To add constraints in the first way, use the ``addConstraint()`` and ``setObjective()``
-functions, e.g.::
+JuMP allows users to use a natural notation to describe linear expressions. To add constraints, use the ``@constraint()`` and ``@objective()``
+macros, e.g.::
 
-    addConstraint(m, y + z == 4)  # Other options: <= and >=
```
Comment (mlubin) [inline doc/quickstart.rst]: What?? This was here?

Comment (joehuchette) [inline doc/quickstart.rst]: Hahaha. Yep....

File: src/deprecated.jl [hunk 4]
```diff
@@ -43,4 +43,46 @@ end
...
+@deprecate JuMPNLPEvaluator JuMP.NLPEvaluator
+@deprecate addConstraint addconstraint
+@deprecate addCutCallback addcutcallback
+@deprecate addHeuristicCallback addheuristiccallback
+@deprecate addInfoCallback addinfocallback
+@deprecate addLazyCallback addlazycallback
+@deprecate addLazyConstraint JuMP.addlazyconstraint
+@deprecate addSolution addsolution
+@deprecate addToExpression JuMP.addtoexpr
+@deprecate addUserCut addusercut
```
Comment (mlubin) [inline src/deprecated.jl]: ``addusercut`` should be unexported as well?

File: doc/quickstart.rst (hunk 3 above)
Comment (IainNZ) [inline doc/quickstart.rst]: lol

File: doc/callbacks.rst [hunk 5]
```diff
@@ -130,10 +130,10 @@ single argument, e.g. ``function myUserCutGenerator(cb)``, where cb is a referen
 to the callback management code inside JuMP. Next you will do whatever
 analysis of the solution you need to inside your function to generate the new
 constraint before adding it to the model with the JuMP function
-``addUserCut(cb, myconstraint)`` or the macro version
-``@addUserCut(cb, myconstraint)`` (same limitations as addConstraint).
+``addusercut(cb, myconstraint)`` or the macro version
```
Comment (IainNZ) [inline doc/callbacks.rst]: I guess this should be removed or changed to `JuMP.addusercut`?

File: doc/nlp.rst [hunk 6]
```diff
@@ -204,7 +204,7 @@ of a JuMP model instead of handing the problem off to a solver.
 Internally, JuMP implements the ``AbstractNLPEvaluator`` interface from
 `MathProgBase <http://mathprogbasejl.readthedocs.org/en/latest/nlp.html>`_.
 To obtain an NLP evaluator object from a JuMP model, use ``JuMPNLPEvaluator``.
-The ``getLinearIndex`` method maps from JuMP variables to the variable
+The ``linearindex`` method maps from JuMP variables to the variable
```
Comment (IainNZ) [inline doc/nlp.rst]: Not sure about this being exported

File: doc/callbacks.rst (hunk 5 above)
Comment (mlubin) [inline doc/callbacks.rst]: removed, I'd say

File: doc/refmodel.rst [hunk 7]
```diff
@@ -47,19 +47,19 @@ Methods
...
 
 **Objective**
 
-* ``getObjective(m::Model)`` - returns the objective function as a ``QuadExpr``.
-* ``setObjective(m::Model, sense::Symbol, a::AffExpr)``, ``setObjective(m::Model, sense::Symbol, q::QuadExpr)`` - sets the objective function to ``a`` and ``q`` respectively, with given objective sense, which must be either ``:Min`` or ``:Max``.
-* ``getObjectiveSense(m::Model)`` - returns objective sense, either ``:Min`` or ``:Max``.
-* ``setObjectiveSense(m::Model, newSense::Symbol)`` - sets the objective sense (``newSense`` is either ``:Min`` or ``:Max``).
-* ``getObjectiveValue(m::Model)`` - returns objective value after a call to ``solve``.
+* ``getobjective(m::Model)`` - returns the objective function as a ``QuadExpr``.
+* ``JuMP.setobjective(m::Model, sense::Symbol, a::AffExpr)``, ``setObjective(m::Model, sense::Symbol, q::QuadExpr)`` - sets the objective function to ``a`` and ``q`` respectively, with given objective sense, which must be either ``:Min`` or ``:Max``.
```
Comment (IainNZ) [inline doc/refmodel.rst]: Should be removed right? Or should be macro version?

Comment (IainNZ) [inline doc/refmodel.rst]: (this isn't a renaming issue though really)

File: doc/callbacks.rst (hunk 5 above)
Comment (joehuchette) [inline doc/callbacks.rst]: should be removed

File: doc/nlp.rst (hunk 6 above)
Comment (joehuchette) [inline doc/nlp.rst]: idk, I find it useful. Not likely to conflict with anything else, either

File: doc/refmodel.rst (hunk 7 above)
Comment (joehuchette) [inline doc/refmodel.rst]: I think it'd probably make sense to just remove it

**Comment by mlubin** [comment]
Travis failures:
```
    Error :: (line:26) :: Dot-ops
      Expression: TestHelper.vec_eq(x ./ A,y ./ A) --> true
      UndefVarError: affstr not defined
       in vec_eq at /Users/travis/.julia/v0.4/JuMP/test/operator.jl:485
       in anonymous at /Users/travis/.julia/v0.4/FactCheck/src/FactCheck.jl:271
       in do_fact at /Users/travis/.julia/v0.4/FactCheck/src/FactCheck.jl:333
       [inlined code] from /Users/travis/.julia/v0.4/FactCheck/src/FactCheck.jl:271
       in anonymous at /Users/travis/.julia/v0.4/JuMP/test/operator.jl:636
       in context at /Users/travis/.julia/v0.4/FactCheck/src/FactCheck.jl:474
       in anonymous at /Users/travis/.julia/v0.4/JuMP/test/operator.jl:589
       in facts at /Users/travis/.julia/v0.4/FactCheck/src/FactCheck.jl:448
       in include at /Users/travis/julia/lib/julia/sys.dylib
       in include_from_node1 at /Users/travis/julia/lib/julia/sys.dylib
       in include at /Users/travis/julia/lib/julia/sys.dylib
       in include_from_node1 at /Users/travis/julia/lib/julia/sys.dylib
       in process_options at /Users/travis/julia/lib/julia/sys.dylib
       in _start at /Users/travis/julia/lib/julia/sys.dylib
```

**Comment by joehuchette** [comment]
OK @mlubin I think this is in decent shape now

**Comment by mlubin** [comment]
@joehuchette, travis still seems to be failing

**Comment by joehuchette** [comment]
AFAICT the travis errors (on v0.4) are something unrelated in GLPK

**Comment by mlubin** [comment]
```
      UndefVarError: cbsetsolutionvalue not defined
       in setsolutionvalue at /Users/travis/.julia/v0.4/JuMP/src/callbacks.jl:238
```
That's in JuMP, should be ``cbsetsolutionvalue!``.

File: doc/refexpr.rst [hunk 8]
```diff
@@ -59,7 +59,7 @@ Methods
 The ``ref`` accepts index sets in the same way as ``@defVar``, and those indices can be used in the construction of the expressions::
 
     @defExpr(m, expr[i=1:3], i*sum{x[j], j=1:3})
-* ``@addSDPConstraint(m::Model, expr)`` - adds a semidefinite constraint to the model ``m``. The expression ``expr`` must be a square, two-dimensional array.
+* ``@SDConstraint(m::Model, expr)`` - adds a semidefinite constraint to the model ``m``. The expression ``expr`` must be a square, two-dimensional array.
```
Comment (joehuchette) [inline doc/refexpr.rst]: ``@SDconstraint``

File: doc/refmodel.rst [hunk 9]
```diff
@@ -47,19 +47,18 @@ Methods
...
 
-* ``getObjective(m::Model)`` - returns the objective function as a ``QuadExpr``.
-* ``setObjective(m::Model, sense::Symbol, a::AffExpr)``, ``setObjective(m::Model, sense::Symbol, q::QuadExpr)`` - sets the objective function to ``a`` and ``q`` respectively, with given objective sense, which must be either ``:Min`` or ``:Max``.
-* ``getObjectiveSense(m::Model)`` - returns objective sense, either ``:Min`` or ``:Max``.
-* ``setObjectiveSense(m::Model, newSense::Symbol)`` - sets the objective sense (``newSense`` is either ``:Min`` or ``:Max``).
-* ``getObjectiveValue(m::Model)`` - returns objective value after a call to ``solve``.
+* ``getobjective(m::Model)`` - returns the objective function as a ``QuadExpr``.
+* ``getobjectivesense(m::Model)`` - returns objective sense, either ``:Min`` or ``:Max``.
+* ``setobjectivesense(m::Model, newSense::Symbol)`` - sets the objective sense (``newSense`` is either ``:Min`` or ``:Max``).
+* ``getobjectivevalue(m::Model)`` - returns objective value after a call to ``solve``.
```
Comment (joehuchette) [inline doc/refmodel.rst]: @mlubin should this be ``objectivevalue`` instead, given our naming convention? (you can't "set" the objective value)

File: doc/refvariable.rst [hunk 10]
```diff
@@ -114,7 +114,7 @@ JuMP supports modeling with `semidefinite variables <https://en.wikipedia.org/wi
 Note in particular the indexing: 1) exactly two index sets must be specified, 2) they must both be unit ranges starting at 1, 3) no bounds can be provided alongside the ``SDP`` tag. If you wish to impose more complex semidefinite constraints on the variables, e.g. :math:`X - I \succeq 0`, you may instead use the ``Symmetric`` tag, along with a semidefinite constraint::
 
     @defVar(m, X[1:n,1:n], Symmetric)
-    @addSDPConstraint(m, X >= eye(n))
+    @SDConstraint(m, X >= eye(n))
```
Comment (joehuchette) [inline doc/refvariable.rst]: ``@SDconstraint``

File: doc/refmodel.rst (hunk 9 above)
Comment (mlubin) [inline doc/refmodel.rst]: It's not really clear to me but might as well go with the shorter name.

File: src/nlp.jl [hunk 11]
```diff
@@ -1172,7 +1172,7 @@ function _buildInternalModel_nlp(m::Model, traits)
             For example:
 
             data = [1.0]
-            @addNLConstraint(m, data[1]*x <= 1)
+            @NLConstraint(m, data[1]*x <= 1)
```
Comment (joehuchette) [inline src/nlp.jl]: ``@NLconstraint``

File: src/print.jl [hunk 12]
```diff
@@ -819,7 +819,7 @@ function con_str(mode, c::GenericRangeConstraint, sym)
     out_str
 end
 # Backwards compatability shim
-conToStr(c::GenericRangeConstraint) = con_str(REPLMode,c)
+constr(c::GenericRangeConstraint) = con_str(REPLMode,c)
```
Comment (joehuchette) [inline src/print.jl]: This reads as "constraint", not "constraint string"...

**Comment by mlubin** [comment]
@chriscoey @tkelman @diam @dpo @davidanthoff @JackDunnNZ @chkwon @anriseth @semartin @evanfields,
This is the 2nd part of the big renaming. All functions are being decamelcased. We opted against a radical approach, e.g., ``setValue`` becomes ``setvalue`` instead of something like ``value!``.
 See [here](https://github.com/JuliaOpt/JuMP.jl/blob/renaming-part-deux/src/deprecated.jl#L48) for the definitive list of proposed changes, which includes unexporting some lesser-used methods. (Unexporting actually lets us use shorter names in some cases, e.g., ``JuMP.register`` versus ``registerNLFunction``.)

Very open to comments, the last discussion was quite helpful in coming to a conclusion on the macro renaming.

File: src/print.jl (hunk 12 above)
Comment (mlubin) [inline src/print.jl]: If these are actually "backwards compatability shims" then we should just be able to remove them by now.

Comment (IainNZ) [inline src/print.jl]: maybe `conToStr` -> `constraintstr`
These "shims" are definitely from me, was some lazyness during the upgrading of printing iirc

**Comment by IainNZ** [comment]
Neat. My eyes are already getting used to it. Except `constr`, that was a mistake.

**Comment by joehuchette** [comment]
Yeah, I'm a fan, despite how massively breaking it is ¯\_(ツ)_/¯

**Comment by diam** [comment]
Le 21 avr. 2016 � 04:43, Miles Lubin <notifications@github.com> a �crit :

> See here for the definitive list of proposed changes, which includes unexporting some lesser-used methods.

I like it too,
-- Maurice

**Comment by anriseth** [comment]
Lowercase looks fine. 
Re: setvalue vs value! The readability of using setvalue, getvalue etc. is good when I present code to colleagues, even if it diverges from the standard Julia-conventions.

**Comment by chkwon** [comment]
Super!

**Comment by chriscoey** [comment]
@joehuchette what happened to your [arm]?(https://www.reddit.com/r/OutOfTheLoop/comments/254y05/why_does_everyone_post_this_%E3%83%84_without_the_arm/)

The new names look better, but I want them to be consistent in terms of naming structure. So if we could categorize the different functions, within each category all the names should the same `noun` or `verbnoun` convention. This is the case for the new names we discussed earlier for variable, constraint etc. But I'm not convinced for the newer changes.

For example, why in the deprecated doc Miles linked to do we have `getObjectiveValue` -> `objectivevalue` but `getValue` -> `getvalue`?

I don't understand what's going on in the deprecated names below [line 44](https://github.com/JuliaOpt/JuMP.jl/blob/renaming-part-deux/src/deprecated.jl#L44). Some of the functions are repeated in the deprecations. Given that this is an important change, I think this file should be commented a little and the functions categorized. @joehuchette, to help this discussion, could you separate these functions categorically (eg, create callback functions, add callback functions, change model functions, get model info functions, print functions, solver functions, internal model functions)? Then we can check that the naming is consistent within each category.

**Comment by mlubin** [comment]
@joehuchette, I'm now thinking that we have to go with ``getobjectivevalue`` for consistency. It's true that there's currently no corresponding ``setobjectivevalue``, but I don't see why there couldn't be in the future, and the distinction in this case isn't obvious to anyone but us.

**Comment by joehuchette** [comment]
I'm alright changing it back to ``getobjectivevalue``.

@chriscoey is [this list of exports](https://github.com/JuliaOpt/JuMP.jl/blob/011a2a56602b52307a63f5cfd638680e8097a535/src/JuMP.jl#L27-L56) more useful for checking that the scheme is consistent? I'm not too worried about unexported names being somewhat inconsistent.

**Comment by chriscoey** [comment]
Thanks Joey that's more helpful. So of the export functions, everything
except linearindex, linearterms, and internalmodel has a verb. Since these
guys aren't changing anything and aren't types, can we stick a get in front
of each? For consistency with the other functions that get (it).

**Comment by mlubin** [comment]
@chriscoey the motivation for dropping the "get" in these cases is that it's not needed for accessing immutable properties, since they can't be "set".

**Comment by chriscoey** [comment]
That's fair too. And the other functions there are all either now or
potentially in future mutable?

**Comment by mlubin** [comment]
That's the idea

**Comment by chriscoey** [comment]
Ok, I'm satisfied.

**Comment by joehuchette** [comment]
Updated with ``getobjectivevalue``.

**Comment by mlubin** [comment]
Only remaining issue is ``constr``?

**Comment by joehuchette** [comment]
Can we just remove them?

**Comment by mlubin** [comment]
If it's easy to do, yes.

**Comment by mlubin** [comment]
What about ``@deprecate conToStr(c::GenericRangeConstraint) JuMP.con_str(JuMP.REPLMode,c)``?

File: doc/quickstart.rst [hunk 13]
```diff
@@ -71,27 +71,20 @@ Finally, bounds can depend on variable indices::
...
 macros, e.g.::
 
-    @addConstraint(m, x[i] - s[i] <= 0)
-    @setObjective(m, Max, sum{x[i], i=1:numLocation} )
+
 
 .. note::
-    The ``sense`` passed to ``setObjective`` must be a `symbol <http://docs.julialang.org/en/latest/manual/metaprogramming/#symbols>`_ type: ``:Min`` or ``:Max``.
-    The ``@setObjective`` macro accepts ``:Min`` and ``:Max``, as well as ``Min`` and ``Max`` (without the colon) directly.
+    The ``sense`` passed to ``@objective`` must be a `symbol <http://docs.julialang.org/en/latest/manual/metaprogramming/#symbols>`_ type: ``:Min`` or ``:Max``, although macro accepts ``:Min`` and ``:Max``, as well as ``Min`` and ``Max`` (without the colon) directly.
```
Comment (mlubin) [inline doc/quickstart.rst]: ``s/although macro/although the macro/``

File: test/macros.jl [hunk 14]
```diff
@@ -203,10 +203,10 @@ facts("[macros] @constraints") do
         ref[i=1:3], y[1] + y[i] >= i
     end)
 
-    @fact conToStr(m.linconstr[1]) --> "x + y[1] $eq 1"
```
Comment (mlubin) [inline test/macros.jl]: Update the use of deprecated ``constr`` in these tests?

Comment (mlubin) [inline test/macros.jl]: There's *a lot* of these. Why don't we have a 1-argument form of ``con_str`` that uses ``REPLMode``?

**Comment by mlubin** [comment]
Woo, green checkmark :100:  :fireworks:

**Comment by IainNZ** [comment]
`addVectorizedConstraint`?

File: doc/quickstart.rst (hunk 3 above)
Comment (tkelman) [inline doc/quickstart.rst]: doctests anyone?

**Comment by IainNZ** [comment]
FYI: `string(c)` works just as well as `JuMP.con_str(JuMP.REPLMode,c)`, as does `string(a)` for `JuMP.aff_str(JuMP.REPLMode,a)`

**Comment by IainNZ** [comment]
Maybe we should change the deprecation message?

**Comment by joehuchette** [comment]
``addVectorizedConstraint`` is unexported and only used internally, so I'm not too worried about changing it up immediately

**Comment by mlubin** [comment]
I'll fix the ``constr`` stuff

**Pull request merged**